### ISSUE TITLE
Modifications made to Capnp.Net.Runtime to reduce GC for low latency applications

### ISSUE THREAD

**Issue by augenix** [c80k/capnproto-dotnetcore#72]
The Framing and FramePump classes use a number of new object allocations for each incoming message.  When message frequency is high (> 100s of thousands of messages per second) these new allocations become very expensive in terms of GC activity.  As is, the runtime is not suitable for low latency applications.

I have modified the Capnp.Net.Runtime classes FramePump and Framing to remove these allocations.  Here is a summary of the modifications.
1. Remove the new BinaryReader allocations.  These are not needed.  Just use the incoming stream and BitConverter.ToInt32() instead. 
2. Remove the new BinaryWriter allocations.  These again are not needed when you have the stream.  
3. Use object pooling approach for the Memory<UInt64> so that these objects are pre-allocated and reused instead of allocated for each message.  Added a struct to the Util namespace to implement the Memory<UInt64> object pool with ConcurrentBag.
4. Use object pooling for the UInt64[] for reuse instead of reallocation.  Added a struct to the Util namespace to implement the UInt64[] object pool with ConcurrentBag.
5. Move various local variables to class level fields where possible so these can be reused.
6. And finally, I updated the runtime to framework target of netapp3.1.  I don't think this is required, but some ideas I tried early in the optimization process would not work with earlier framework targets.  I suppose it could be reverted back and still work.

The result of these modifications have reduced GC to a level that is in line with the requirements of low latency applications.

I look forward to comments and questions.

**Comment by c80k** [comment]
Sounds promising! What about a PR?

**Comment by augenix** [comment]
I am working towards a PR, but this will probably end up as a breaking change.  The fundamental change that ripples through everything is to use Memory< byte > to hold the entire message (including all segments) as a flat, contiguous block of memory to replace the IReadOnlyList<Memory< ulong >> WireFrame object to hold each segment in a list element.   The underlying byte[] buffer for the Memory< byte > type is pre-allocated in the client application and returned to the client buffer allocation pool after use.   However, IReadOnlyList and ulong are managed memory types and must be allocated to the heap for each message.  This creates significant GC pressure. 

Fortunately, the existing runtime is using Span of Memory< ulong > segment elements so this is somewhat compatible.   Moving to Span of Memory< byte > requires that indexing be managed manually using Slice(start, length).

I am still working through the remaining issues where Memory< byte > breaks the code.  Then there are still numerous 'new' object allocations that go onto the heap that must be somehow eliminated. 

I recompiled everything now using .NET 6.0 and C#10 and removed all of the compiler conditionals.  

I am only using CapnProto as a serialization protocol for my API.  I am not using RPC.  Therefore, I probably will not update the RPC code and simply remove it from the Runtime (if possible).

Any advice is appreciated.  If it makes sense to do a PR after all of this I will.

**Comment by augenix** [comment]
It appears that the scope of changes to make Runtime allocation free are vastly more significant than originally estimated. Moreover, the changes are not surgical tweaks.  A sledge hammer is required.  The effort to revise the codebase appears to be substantially higher than the effort to create the functionality that my API needs from scratch.  I have decided not to do a pull request at this time.  I am going to leave this issue open for a couple of weeks for comments.  After that it can be closed.

**Comment by augenix** [comment]
The work that C80k has done to build this Runtime is phenomenal.  Attempts to modify the codebase to remove allocations was really difficult and I came to appreciate more the work that had been done.  I abandoned trying to modify the codebase and solved the allocation issues with object preallocation and recovery techniques in the application layer.  I did not modify the original code.  

I did create a specific implementation of Framing.ReadFrameSegments() that worked for my specific needs.  It is similar to the original with two major differences.  First, the Memory < ulong >[] and the ulong[] are preallocated so they are pulled from a pool instead of being instantiated.  Second,  the algo uses a stateful approach that doesn't block during attempts to read the stream.  The result is something similar to asynchronous IO, but with higher determinism.  

Closing this issue.
